store/mongo/datastore_mongo.go

```diff
@@ -360,38 +422,110 @@ func (db *DataStoreMongo) GetSessionRecording(ctx context.Context,
 	if err != nil {
 		return err
 	}
+	defer recordingsCursor.Close(ctx)
 
-	output := make([]byte, recordingReadBufferSize)
-	var buffer bytes.Buffer
-	for c.Next(ctx) {
-		var r model.Recording
-		err = c.Decode(&r)
-		if err != nil {
-			return err
-		}
-
-		buffer.Reset()
-		buffer.Write(r.Recording)
-		gzipReader, e := gzip.NewReader(&buffer)
-		if e != nil {
-			err = e
-		}
-
-		for {
-			n, e := gzipReader.Read(output)
-			if n == 0 || e != nil {
-				gzipReader.Close()
+	controlCursor, err := collControl.Find(ctx,
+		bson.M{
+			dbFieldSessionID: sessionID,
+		},
+		findOptions,
+	)
+	if err != nil {
+		return err
+	}
+	defer controlCursor.Close(ctx)
+
+	controlReader := NewControlMessageReader(ctx, controlCursor)
+	recordingReader := NewRecordingReader(ctx, recordingsCursor)
+
+	recordingBuffer := make([]byte, recordingReadBufferSize)
+	recordingBytesSent := 0
+	for {
+		control := controlReader.Pop()
+		if control == nil {
+			l.Debug("WriteSessionRecords: no more control " +
+				"messages, flushing the recording upstream.")
+			//no more control messages, we send the whole recording
+			var err error = nil
+			var n int
+			for err != io.EOF {
+				n, err = recordingReader.Read(recordingBuffer)
+				if n > 0 {
+					//nolint:errcheck
+					sendRecordingMessage(recordingBuffer[:n], sessionID, w)
+					recordingBytesSent += n
+				}
+			}
+			break
+		} else {
```

Comment:
This case could be simplified a lot more by using `io.CopyN` with error handling (requires adding a Writer interface to wrap `sendRecordingMessage`).

Reply:
sure thing. added RecordingWriter